File: server/services/menuService.js
const axios = require('axios');

/**
 * 生成菜单服务
 */
class MenuService {
  constructor() {
    this.apiProvider = process.env.AI_PROVIDER || 'glm';
  }

  /**
   * 调用AI API生成菜单
   */
  async generateMenu({ mealType, preferences, servings, restrictions }) {
    const prompt = this.buildPrompt(mealType, preferences, servings, restrictions);
    
    try {
      // 根据配置选择AI提供商
      let menu;
      switch (this.apiProvider) {
        case 'glm':
          menu = await this.callGLMAPI(prompt);
          break;
        case 'openai':
          menu = await this.callOpenAI(prompt);
          break;
        case 'claude':
          menu = await this.callClaude(prompt);
          break;
        default:
          menu = await this.callGLMAPI(prompt);
      }

      return menu;
    } catch (error) {
      console.error('AI API调用失败:', error);
      // 返回模拟数据作为降级方案
      return this.getMockMenu(mealType);
    }
  }

  /**
   * 构建提示词
   */
  buildPrompt(mealType, preferences, servings, restrictions) {
    let prompt = `请为${servings}人生成一份${mealType}菜单，`;
    
    if (preferences.length > 0) {
      prompt += `口味偏好：${preferences.join('、')}，`;
    }
    
    if (restrictions.length > 0) {
      prompt += `饮食限制：${restrictions.join('、')}，`;
    }
    
    prompt += `\n请以JSON格式返回，包含以下字段：
- dishes: 菜品数组，每个菜品包含 name(菜名), description(描述), difficulty(难度1-5), time(预计时间)
- nutrition: 营养信息对象，包含 calories, protein, carbs, fat`;

    return prompt;
  }

  /**
   * 调用GLM API
   */
  async callGLMAPI(prompt) {
    const response = await axios.post('https://open.bigmodel.cn/api/paas/v3/model-api/chatglm_pro/invoke', {
      prompt: prompt,
      temperature: 0.7,
      top_p: 0.9
    }, {
      headers: {
        'Authorization': `Bearer ${process.env.GLM_API_KEY}`,
        'Content-Type': 'application/json'
      }
    });

    return this.parseAIResponse(response.data.data.choices[0].content);
  }

  /**
   * 调用OpenAI API
   */
  async callOpenAI(prompt) {
    const response = await axios.post('https://api.openai.com/v1/chat/completions', {
      model: 'gpt-3.5-turbo',
      messages: [{ role: 'user', content: prompt }],
      temperature: 0.7
    }, {
      headers: {
        'Authorization': `Bearer ${process.env.OPENAI_API_KEY}`,
        'Content-Type': 'application/json'
      }
    });

    return this.parseAIResponse(response.data.choices[0].message.content);
  }

  /**
   * 调用Claude API
   */
  async callClaude(prompt) {
    const response = await axios.post('https://api.anthropic.com/v1/messages', {
      model: 'claude-3-sonnet-20240229',
      max_tokens: 1024,
      messages: [{ role: 'user', content: prompt }]
    }, {
      headers: {
        'x-api-key': process.env.CLAUDE_API_KEY,
        'anthropic-version': '2023-06-01',
        'Content-Type': 'application/json'
      }
    });

    return this.parseAIResponse(response.data.content[0].text);
  }

  /**
   * 解析AI响应
   */
  parseAIResponse(content) {
    try {
      // 尝试提取JSON
      const jsonMatch = content.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        return JSON.parse(jsonMatch[0]);
      }
    } catch (error) {
      console.error('解析AI响应失败:', error);
    }
    
    // 如果解析失败，返回模拟数据
    return this.getMockMenu('午餐');
  }

  /**
   * 模拟数据（降级方案）
   */
  getMockMenu(mealType) {
    const menus = {
      '早餐': {
        dishes: [
          { name: '鸡蛋灌饼', description: '酥脆的饼皮包裹嫩滑鸡蛋', difficulty: 2, time: '15分钟' },
          { name: '小米粥', description: '养胃暖身，搭配小菜更美味', difficulty: 1, time: '30分钟' },
          { name: '凉拌黄瓜', description: '清爽开胃，简单快手', difficulty: 1, time: '5分钟' }
        ],
        nutrition: { calories: 450, protein: 15, carbs: 60, fat: 18 }
      },
      '午餐': {
        dishes: [
          { name: '红烧肉', description: '肥而不腻，入口即化', difficulty: 4, time: '90分钟' },
          { name: '清炒时蔬', description: '新鲜蔬菜，清淡爽口', difficulty: 2, time: '10分钟' },
          { name: '番茄蛋汤', description: '酸甜开胃，营养丰富', difficulty: 2, time: '15分钟' }
        ],
        nutrition: { calories: 650, protein: 25, carbs: 45, fat: 35 }
      },
      '晚餐': {
        dishes: [
          { name: '清蒸鲈鱼', description: '鲜嫩多汁，保留原味', difficulty: 3, time: '25分钟' },
          { name: '蒜蓉西兰花', description: '营养健康，色彩诱人', difficulty: 2, time: '10分钟' },
          { name: '紫菜蛋花汤', description: '清淡鲜美，助消化', difficulty: 1, time: '10分钟' }
        ],
        nutrition: { calories: 400, protein: 30, carbs: 25, fat: 15 }
      }
    };

    return menus[mealType] || menus['午餐'];
  }

  /**
   * 获取推荐菜品
   */
  async getRecommendations() {
    return {
      popular: [
        { name: '宫保鸡丁', rating: 4.8, orders: 1256 },
        { name: '麻婆豆腐', rating: 4.7, orders: 987 },
        { name: '糖醋里脊', rating: 4.9, orders: 1532 }
      ],
      seasonal: [
        { name: '春笋炒肉', reason: '春季时令' },
        { name: '韭菜盒子', reason: '春季养生' }
      ]
    };
  }
}

module.exports = new MenuService();

File: server-rust/src/services/menu_service.rs
use crate::error::AppError;
use crate::models::Menu;
use reqwest::Client;
use serde_json::json;

pub struct MenuService {
    client: Client,
    ai_provider: String,
    glm_api_key: Option<String>,
    openai_api_key: Option<String>,
}

impl MenuService {
    pub fn new(
        ai_provider: String,
        glm_api_key: Option<String>,
        openai_api_key: Option<String>,
    ) -> Self {
        Self {
            client: Client::new(),
            ai_provider,
            glm_api_key,
            openai_api_key,
        }
    }

    pub async fn generate_menu(
        &self,
        meal_type: &str,
        servings: u8,
        preferences: &[String],
        restrictions: &[String],
    ) -> Result<Menu, AppError> {
        // 尝试使用 AI 生成
        if let Some(menu) = self
            .try_ai_generation(meal_type, servings, preferences, restrictions)
            .await?
        {
            return Ok(menu);
        }

        // 降级到模拟数据
        Ok(self.get_mock_menu(meal_type, servings))
    }

    async fn try_ai_generation(
        &self,
        meal_type: &str,
        servings: u8,
        preferences: &[String],
        restrictions: &[String],
    ) -> Result<Option<Menu>, AppError> {
        match self.ai_provider.as_str() {
            "glm" => {
                if let Some(api_key) = &self.glm_api_key {
                    match self
                        .call_glm_api(meal_type, servings, preferences, restrictions, api_key)
                        .await
                    {
                        Ok(menu) => Ok(Some(menu)),
                        Err(e) => {
                            tracing::warn!("GLM API 调用失败: {}, 使用降级方案", e);
                            Ok(None)
                        }
                    }
                } else {
                    Ok(None)
                }
            }
            "openai" => {
                if let Some(api_key) = &self.openai_api_key {
                    match self
                        .call_openai_api(meal_type, servings, preferences, restrictions, api_key)
                        .await
                    {
                        Ok(menu) => Ok(Some(menu)),
                        Err(e) => {
                            tracing::warn!("OpenAI API 调用失败: {}, 使用降级方案", e);
                            Ok(None)
                        }
                    }
                } else {
                    Ok(None)
                }
            }
            _ => Ok(None),
        }
    }

    async fn call_glm_api(
        &self,
        meal_type: &str,
        servings: u8,
        preferences: &[String],
        restrictions: &[String],
        api_key: &str,
    ) -> Result<Menu, AppError> {
        let prompt = format!(
            "请为{}生成{}人份的菜单。饮食偏好：{}。饮食限制：{}。",
            meal_type,
            servings,
            preferences.join("、"),
            restrictions.join("、")
        );

        let _response = self
            .client
            .post("https://open.bigmodel.cn/api/paas/v3/model-api/chatglm_pro/invoke")
            .header("Authorization", format!("Bearer {}", api_key))
            .json(&json!({
                "prompt": prompt,
                "temperature": 0.7,
            }))
            .send()
            .await
            .map_err(|e| AppError::AIServiceError(format!("GLM API 请求失败: {}", e)))?;

        // TODO: 实现完整的解析逻辑
        Err(AppError::AIServiceError(
            "GLM 响应解析未完成".to_string(),
        ))
    }

    async fn call_openai_api(
        &self,
        meal_type: &str,
        servings: u8,
        preferences: &[String],
        restrictions: &[String],
        api_key: &str,
    ) -> Result<Menu, AppError> {
        let prompt = format!(
            "请为{}生成{}人份的菜单。饮食偏好：{}。饮食限制：{}。返回 JSON 格式。",
            meal_type,
            servings,
            preferences.join("、"),
            restrictions.join("、")
        );

        let _response = self
            .client
            .post("https://api.openai.com/v1/chat/completions")
            .header("Authorization", format!("Bearer {}", api_key))
            .json(&json!({
                "model": "gpt-3.5-turbo",
                "messages": [
                    {"role": "system", "content": "你是一个专业的营养师和厨师助手。"},
                    {"role": "user", "content": prompt}
                ],
                "temperature": 0.7,
            }))
            .send()
            .await
            .map_err(|e| AppError::AIServiceError(format!("OpenAI API 请求失败: {}", e)))?;

        // TODO: 实现完整的解析逻辑
        Err(AppError::AIServiceError(
            "OpenAI 响应解析未完成".to_string(),
        ))
    }

    fn get_mock_menu(&self, meal_type: &str, servings: u8) -> Menu {
        let dishes = match meal_type {
            "早餐" => vec![
                crate::models::Dish {
                    name: "小米粥".to_string(),
                    description: "养胃健脾，营养丰富".to_string(),
                    difficulty: 1,
                    time: "30分钟".to_string(),
                    calories: Some(150 * servings as u32),
                },
                crate::models::Dish {
                    name: "鸡蛋饼".to_string(),
                    description: "简单快手，美味可口".to_string(),
                    difficulty: 2,
                    time: "15分钟".to_string(),
                    calories: Some(200 * servings as u32),
                },
            ],
            "午餐" => vec![
                crate::models::Dish {
                    name: "红烧肉".to_string(),
                    description: "经典家常菜，肥而不腻".to_string(),
                    difficulty: 4,
                    time: "90分钟".to_string(),
                    calories: Some(520 * servings as u32),
                },
                crate::models::Dish {
                    name: "清炒时蔬".to_string(),
                    description: "清淡爽口，营养健康".to_string(),
                    difficulty: 2,
                    time: "10分钟".to_string(),
                    calories: Some(80 * servings as u32),
                },
            ],
            "晚餐" => vec![
                crate::models::Dish {
                    name: "清蒸鲈鱼".to_string(),
                    description: "鲜嫩多汁，保留原味".to_string(),
                    difficulty: 3,
                    time: "25分钟".to_string(),
                    calories: Some(180 * servings as u32),
                },
                crate::models::Dish {
                    name: "番茄蛋汤".to_string(),
                    description: "酸甜可口，开胃消食".to_string(),
                    difficulty: 1,
                    time: "15分钟".to_string(),
                    calories: Some(120 * servings as u32),
                },
            ],
            _ => vec![crate::models::Dish {
                name: "营养套餐".to_string(),
                description: "均衡营养，健康美味".to_string(),
                difficulty: 2,
                time: "30分钟".to_string(),
                calories: Some(300 * servings as u32),
            }],
        };

        Menu {
            meal_type: Some(meal_type.to_string()),
            servings: Some(servings),
            dishes,
            nutrition: Some(crate::models::MenuNutrition {
                total_calories: 800 * servings as u32,
                protein: 35 * servings as u32,
                carbs: 100 * servings as u32,
                fat: 25 * servings as u32,
            }),
        }
    }

    pub async fn get_recommendations() -> Menu {
        Menu {
            meal_type: Some("推荐".to_string()),
            servings: Some(2),
            dishes: vec![
                crate::models::Dish {
                    name: "红烧肉".to_string(),
                    description: "经典家常菜，肥而不腻".to_string(),
                    difficulty: 4,
                    time: "90分钟".to_string(),
                    calories: Some(520),
                },
                crate::models::Dish {
                    name: "清蒸鲈鱼".to_string(),
                    description: "鲜嫩多汁，保留原味".to_string(),
                    difficulty: 3,
                    time: "25分钟".to_string(),
                    calories: Some(180),
                },
                crate::models::Dish {
                    name: "宫保鸡丁".to_string(),
                    description: "酸甜微辣，经典川菜".to_string(),
                    difficulty: 3,
                    time: "30分钟".to_string(),
                    calories: Some(380),
                },
            ],
            nutrition: Some(crate::models::MenuNutrition {
                total_calories: 1080,
                protein: 65,
                carbs: 45,
                fat: 72,
            }),
        }
    }
}
